Add well-known Debug.Assert method

jcouv opened the pull request:
`Assert.Debug(condition)` is analyzed as `if (!condition) throw;`.

CyrusNajmabadi commented:
I feel like it would be better if these methods could somehow be annotated.  Or if this info could be plugged intot he compiler through some extensibility mechanism.  Otherwise, this is just going to be wackamole.

jcouv commented:
@CyrusNajmabadi We're planning to do a little of both.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1577,6 +1595,66 @@ public override BoundNode VisitCall(BoundCall node)
+            return method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_Diagnostics_Debug_Assert1)) ||
+                method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_Diagnostics_Debug_Assert2)) ||
+                method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_Diagnostics_Debug_Assert3)) ||
+                method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_Diagnostics_Debug_Assert4));
```
Why use `WellKnownMember` rather than the extra annotations table? (Presumably, we won't need these well-known members once we support annotations.)

jcouv replied:
What we discussed in LDM was that some members would be well-known. That said, maybe that was only for built-in operators?
I like the extra annotations approach. I'll try that.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1568,6 +1574,18 @@ public override BoundNode VisitCall(BoundCall node)
+            {
+                VisitTrueWhenNotNull(arguments[0], method.ReturnType.TypeSymbol);
+                _result = method.ReturnType;
+                return null;
```
Are the last two statements necessary?

jcouv replied:
We need to return, since the assertion below would be violated. 
We need to set the result, otherwise another assertion breaks (we check that the result is not left as invalid type).

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1577,6 +1595,66 @@ public override BoundNode VisitCall(BoundCall node)
+                new BoundBinaryOperator(operand.Syntax, BinaryOperatorKind.Equal, operand,
+                    new BoundLiteral(operand.Syntax, ConstantValue.Null, operand.Type),
+                    constantValueOpt: null, methodOpt: null, LookupResultKind.Viable, boolType),
+                new BoundLiteral(operand.Syntax, constantValueOpt: null, boolType));
```
We're tracking "maybe null" rather than "definitely null". In short, is the `|| someUnknownBoolean` necessary?

cston replied:
Can this be implemented without synthesizing nodes?

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1577,6 +1595,66 @@ public override BoundNode VisitCall(BoundCall node)
+
+            // if (!condition) throw;
+            VisitIfStatement(new BoundIfStatement(condition.Syntax, SyntheticBoundNodeFactory.NotCore(condition),
+                new BoundThrowStatement(condition.Syntax, expressionOpt: null), alternativeOpt: null));
```
Can we implement this visit method without synthesizing nodes, by mimicking the behavior of `VisitIfStatement` and `VisitThrowExpression`?

jcouv replied:
I didn't want to duplicate that logic, but we can discuss some alternatives.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1577,6 +1595,66 @@ public override BoundNode VisitCall(BoundCall node)
+                return false;
+            }
+
+            return method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_String_IsNullOrEmpty));
```
Please add a `// PROTOTYPE` comment to replace this with checking annotations. Same for `MethodEnsuresTrueWhenExits`.

cston commented on src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs:
```diff
@@ -1252,8 +1252,12 @@ internal static BoundExpression NullOrDefault(TypeSymbol typeSymbol, SyntaxNode
+            return NotCore(expression);
+        }
+
+        internal static BoundUnaryOperator NotCore(BoundExpression expression)
```
Perhaps change the return type of the original method.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1568,6 +1574,18 @@ public override BoundNode VisitCall(BoundCall node)
 
+            if (MethodEnsuresTrueWhenExits(method) && arguments.Length > 0)
+            {
+                VisitTrueWhenExits(condition: arguments[0]);
```
The beginning of `VisitCall` has a check for `method.CallsAreOmitted` with a comment to skip the visitor in the future when the `[Conditional]` attribute is not satisfied. But for methods such as `Debug.Assert` (and similar methods), it seems we should visit the condition argument regardless of whether we visit other arguments. Consider having one test with both `TestOptions.DebugDll` and `TestOptions.ReleaseDll` to ensure that.

jcouv replied:
Added test `EnsuresNotNull_ConditionalMethodInReleaseMode`

---
In reply to: [186272483](https://github.com/dotnet/roslyn/pull/26656#discussion_r186272483) [](ancestors = 186272483)

sharwell commented:
> Debug.Assert

This one is interesting. Do you have examples of where it impacts code?

AArnott commented:
The syntax I use was inspired by powershell, incidentally.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -11,6 +11,7 @@
 using System.Diagnostics;
 using System.Linq;
 using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE;
```
Perhaps move `ExternalAnnotations` to `Microsoft.CodeAnalysis.CSharp.Symbols`.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1577,6 +1606,97 @@ public override BoundNode VisitCall(BoundCall node)
             return null;
         }
 
+        private BoundExpression TryGetFirstArgument(MethodSymbol method, ImmutableArray<BoundExpression> arguments)
```
`static`

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1577,6 +1606,97 @@ public override BoundNode VisitCall(BoundCall node)
+                return (arguments.Length > 1) ? arguments[1] : null;
+            }
+
+            return (arguments.Length > 0) ? arguments[0] : null;
```
Perhaps calculate index separately from `return arguments.Length > index ? ...;`

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1577,6 +1606,97 @@ public override BoundNode VisitCall(BoundCall node)
+            return (arguments.Length > 0) ? arguments[0] : null;
+        }
+
+        private bool MethodReturnsTrueWhenNotNull(MethodSymbol method, ImmutableArray<AttributeDescription> extraAttributes)
```
`static`

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1577,6 +1606,97 @@ public override BoundNode VisitCall(BoundCall node)
+            return MethodHasAttribute(method, extraAttributes, AttributeDescription.EnsuresTrueWhenExitsAttribute);
+        }
+
+        private bool MethodHasAttribute(MethodSymbol method, ImmutableArray<AttributeDescription> extraAttributes, AttributeDescription description)
```
`static`

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1577,6 +1606,97 @@ public override BoundNode VisitCall(BoundCall node)
+                {
+                    return true;
+                }
+            }
```
Should this loop be skipped if `extraAttributes` is set?

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1577,6 +1606,97 @@ public override BoundNode VisitCall(BoundCall node)
+                    {
+                        return true;
+                    }
+                }
```
`return extraAttributes.Contains(description);`

cston commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/ExternalAnnotations.cs:
```diff
@@ -96,5 +145,15 @@ private static void Add(TypeSymbol type, StringBuilder builder)
+
+        internal static ImmutableArray<AttributeDescription> GetExtraAttributes(string key)
+        {
+            if (!Attributes.TryGetValue(key, out var descriptions))
```
The `if` is not necessary. `return descriptions;` should work regardless of result of `TryGetValue`.

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:
```diff
@@ -5437,6 +5462,549 @@ class CL1<T>
+using System.Runtime.CompilerServices;
+class C
+{
+    void Main(C? c)
```
`C c` instead?

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:
```diff
@@ -5437,6 +5462,549 @@ class CL1<T>
+    void Main(C c)
+    {
+        System.Diagnostics.Debug.Assert(c == null, ""hello"");
+        c.ToString();
```
`warning CS8602: Possible dereference of a null reference`?

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:
```diff
@@ -5437,6 +5462,549 @@ class CL1<T>
+        }
+
+        [Fact]
+        public void EnsuresTrueWhenExits_Debug_Assert_NoDuplicateDiagnostics()
```
What are we testing? Perhaps add a comment.

cston commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/ExternalAnnotations.cs:
```diff
@@ -31,6 +31,17 @@ internal static class ExtraAnnotations
+                { "System.Void System.Diagnostics.Debug.Assert(System.Boolean, System.String)", Descriptions(AttributeDescription.EnsuresTrueWhenExitsAttribute) },
+                { "System.Void System.Diagnostics.Debug.Assert(System.Boolean, System.String, System.String)", Descriptions(AttributeDescription.EnsuresTrueWhenExitsAttribute) },
+                { "System.Void System.Diagnostics.Debug.Assert(System.Boolean, System.String, System.String, System.Object[])", Descriptions(AttributeDescription.EnsuresTrueWhenExitsAttribute) },
+                { "System.Boolean System.String.IsNullOrEmpty(System.String)", Descriptions(AttributeDescription.TrueWhenNotNullAttribute) },
```
`FalseWhenNotNullAttribute`?

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1577,6 +1606,97 @@ public override BoundNode VisitCall(BoundCall node)
+
+        private void VisitTrueWhenNotNull(BoundExpression operand, TypeSymbol boolType)
+        {
+            if (!operand.Type.IsReferenceType)
```
Add test for `IsNullOrEmpty(null)`.

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:
```diff
@@ -5437,6 +5462,549 @@ class CL1<T>
+        }
+
+        [Fact]
+        public void TrueWhenNotNull_String_NotIsNullOrEmpty()
```
Already covered by `TrueWhenNotNull_InvertIf`.

jcouv replied:
It is indeed a bit redundant, but there is a difference. This test checks the BCL API, whereas `TrueWhenNotNull_InvertIf` checks a source method.

---
In reply to: [186469012](https://github.com/dotnet/roslyn/pull/26656#discussion_r186469012) [](ancestors = 186469012)

cston commented on src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs:
```diff
@@ -28,6 +28,12 @@ public AttributeDescription(string @namespace, string name, byte[][] signatures,
+            => this.Namespace == other.Namespace &&
+                this.Name == other.Name &&
+                ReferenceEquals(this.Signatures, other.Signatures) &&
+                this.MatchIgnoringCase == other.MatchIgnoringCase;
```
Consider moving this to `NullableWalker.MethodHasAttribute` since the method is only used for that (temporary?) scenario.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1577,6 +1606,97 @@ public override BoundNode VisitCall(BoundCall node)
+            _disableDiagnostics = true;
+
+            // if (!condition) throw;
+            VisitIfStatement(new BoundIfStatement(condition.Syntax, SyntheticBoundNodeFactory.NotCore(condition),
```
>new BoundIfStatement(condition.Syntax, SyntheticBoundNodeFactory.NotCore(condition) [](start = 29, length = 83)

I think generating bound nodes in this walker is the wrong thing to do.

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs:
```diff
@@ -28,6 +28,12 @@ public AttributeDescription(string @namespace, string name, byte[][] signatures,
             this.MatchIgnoringCase = matchIgnoringCase;
         }
 
+        public bool Equals(AttributeDescription other)
```
>public bool Equals(AttributeDescription other) [](start = 8, length = 46)

Why is this method needed? Implementing equality that works well for everybody is not a trivial thing. If you need some kind of comparison for the purposes of specific component, it is much better to define a custom comparer locally.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1568,6 +1575,28 @@ public override BoundNode VisitCall(BoundCall node)
                 ReplayReadsAndWrites(localFunc, node.Syntax, writes: true);
             }
 
+            string key = ExtraAnnotations.MakeMethodKey(method);
```
>string key = ExtraAnnotations.MakeMethodKey(method); [](start = 12, length = 52)

I think all annotations should come from symbols, this component shouldn't even know that such thing as ExtraAnnotations exist.

jcouv commented:
From discussion with Chuck, Aleskey and Neal, I'm going to modify this PR with the following design:
- `NullableWalker.VisitCall` won't synthesize "equivalent" bound nodes, but rather directly manipulate the visitor's state,
- I'll add properties to `MethodSymbol` to indicate the two special behaviors, and those properties will be driven by the presence of attributes and extra attributes.

Also, let me take a step back and describe the overall design for this PR:
We want to allow some attributes, placed on methods (or maybe on the parameters?), to affect the nullable flow analysis. For instance, we should know that `x` isn't `null` after `Debug.Assert(x != null);` (assuming that `Debug.Assert` has the proper attribute).
So there is a change to `NullableWalker.VisitCall` to handle that.

Then, there is a second part of this change, which is that we want this same behavior on some BCL methods, before we can ship an updated BCL with proper attributes. Although the overall design for such external annotations is not finalized, we have an abstraction (`ExtraAnnotations`) that lets you ask for nullable annotations for a given method (keyed by some format). This PR adds the ability to ask for attribute annotations for a given method.

gafter commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1577,6 +1606,97 @@ public override BoundNode VisitCall(BoundCall node)
+        private BoundExpression TryGetFirstArgument(MethodSymbol method, ImmutableArray<BoundExpression> arguments)
+        {
+            if (method.IsExtensionMethod)
+            {
```
This does not seem correct. An extension method can still be invoked as an ordinary static method. Perhaps you are trying to check if it was invoked *as* an extension method? Or maybe this is really what you intend (check argument 0 of the method, or argument 1 if the first parameter has a `this` modifier)?

jcouv commented:
Note: From Chuck's notes from LDM, I'm going to do one more design change: using attributes on parameters, instead of on methods.

```C#
static bool IsNullOrEmpty([NotNullWhenFalse] string? s) { }  
static void AssertNotNull<T>([EnsuresNotNull] T? t) where T : class { }  
```

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:
```diff
@@ -5437,6 +5462,549 @@ class CL1<T>
+        }
+
+        [Fact]
+        public void EnsuresTrueWhenExits_Debug_Assert2_NotNull()
```
Could refactor a lot of these tests with no diagnostics into a single theory with a nullability (for the parameter) and a set of arguments.

jcouv replied:
Most tests now include some additional validation. Not many tests remain that only verify empty diagnostics. I'll leave those couple ones as-is.

---
In reply to: [186580108](https://github.com/dotnet/roslyn/pull/26656#discussion_r186580108) [](ancestors = 186580108)

jcouv commented:
I've made a few major adjustments to this PR:
- removed the attribute for `Debug.Assert` (that will be a separate PR)
- moved the detection of attributes or external annotations into parameter symbols
- the nullable analysis in presence of those attributes is directly manipulating state (rather than generating bound nodes)

I have some work left in this PR, but I believe it's useful to validate I'm on the right track.
I still intend to add validation logic in parameter symbols to warn if attributes are improperly used in source, and ignore them if improperly used in metadata. 
I also intend to add more tests (symbols from metadata, object creation, attribute on `this` parameter in extension method).

I've found some tricky scenarios however. I'll describe them below with my proposed solution.

Let's assume `void M([EnsuresNotNull] string? s, string? s2)`.
If you invoke it with `M(s, s.ToString())`, then you want a warning on the second argument, but you want `s` to be not-null after the method returns.
If you invoke it with `M(s, s = null)`, then you want `s` to be maybe-null after the method returns.

The solution is to analyze the arguments in this invocation twice:
- The first time is like today. It will flow state from one argument evaluation to the next, and warn as needed on the second argument.
- The second time will silently evaluate the arguments again (dropping any diagnostics), but it honors the attributes. So `s` is considered not-null after the first argument evaluation (because the parameter is marked with `[EnsuresNotNull]`).

For `[NotNullWhenFalse]` there is a similar problem. Let's assume `bool M([NotNullWhenFalse] string? s, string? s2)`.
If you invoke it with `if (M(s, s.ToString()) ...`, then you want a warning on the second argument, but `s` is not-null in one of the branches of the `if` statement.
If you invoked it with `if (M(s, s = null)) ...`, then you want `s` to be maybe-null in both branches.

The same design works, except that the second pass on the arguments will track two states independently (one for the invocation returning `true` and one for the invocation return `false`).

I believe this design will also work with `[EnsuresTrue]` and `[EnsuresFalse]`. To achieve that, during the second pass, we'll evaluate such arguments in such a way that we don't collapse their conditional states and we will set the state to be either the `WhenTrue` or the `WhenFalse` state (depending on the attribute present).

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1644,6 +1685,25 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
+
+            // We do a second pass through the arguments, ignoring any diagnostics produced, but honoring the annotations,
+            // to get the proper result state.
+            this.SetState(savedState);
```
:memo: I suspect there is no need to reset state between the two passes, because:
- it's only possible for the second pass to produce a more restrictive state (more variables are considered non-null),
- starting from an incorrect state would affect diagnostics, but they are silenced in the second pass.

cston commented on src/Compilers/CSharp/Portable/Symbols/ExternalAnnotations.cs:
```diff
@@ -0,0 +1,174 @@
+    // PROTOTYPE(NullableReferenceTypes): external annotations should be removed or fully designed/productized
+    //  If we choose to stick with an ad-hoc key (rather than annotations as source or as PE/ref assembly),
+    //  we should consider the assembly qualified name format used in metadata (with backticks and such).
+    internal static class ExtraAnnotations
```
Consider using same name for class and file.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1577,6 +1587,37 @@ public override BoundNode VisitCall(BoundCall node)
+        /// For each argument, figure out if its corresponding parameter is annotated with NotNullWhenFalse or
+        /// EnsuresNotNull.
+        /// </summary>
+        private ImmutableArray<(bool notNullWhenFalse, bool ensuresNotNull)> GetAnnotations(int numArguments,
```
Please consider using uppercase for first letter of tuple element names.

cston replied:
Consider using an `enum` value rather than two separate `bool`s.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1658,22 +1718,134 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
+                    slots.Add(null);
+                    continue;
+                }
+                if (slot >= this.State.Capacity) Normalize(ref this.State);
```
Minor point: It should be sufficient to call `Normalize(ref this.State);` once, unconditionally, after the loop.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1658,22 +1718,134 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
+                    VisitArgumentEvaluate(arguments, refKindsOpt, i);
+                    Debug.Assert(!IsConditionalState);
+
+                    this.SetConditionalState(whenTrue, whenFalse);
```
`LocalState` is a `struct` in `RELEASE` build, so it looks like `whenTrue` and `whenFalse` need to be reassigned after each corresponding call to `VisitArgumentEvaluate`.

jcouv replied:
That would explain a test failure in CI release-mode. Thanks!

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1658,22 +1718,134 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
+                }
+
+                var argument = arguments[i];
+                if (argument.Type?.IsReferenceType != true || slots[i] == null)
```
We may need to handle unconstrained type parameters as well: `AssertNotNull<T>([EnsuresNotNull] T t);`. Perhaps add a `// PROTOTYPE` comment.

jcouv replied:
I'll be adding a test (`EnsuresNotNull_GenericMethod`). It just worked though. Lucky

---
In reply to: [187071624](https://github.com/dotnet/roslyn/pull/26656#discussion_r187071624) [](ancestors = 187071624)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1658,22 +1718,134 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
+                int slot = MakeSlot(argument);
+                if (slot <= 0)
+                {
+                    slots.Add(null);
```
Consider using `slot` as is, for consistency with the rest of the class, rather than using `null` to represent no slot.

cston commented on src/Compilers/CSharp/Portable/Symbols/ParameterSymbol.cs:
```diff
@@ -385,6 +385,56 @@ internal sealed override ObsoleteAttributeData ObsoleteAttributeData
+        /// <summary>
+        /// When the method returns, then we can infer that the argument used for this parameter was not null.
+        /// </summary>
+        internal abstract bool EnsuresNotNull { get; }
```
Should these two properties be combined, with a single `enum` value?
```
[Flags] enum EnsuresNullability { None = 0, NotNullWhenFalse = 1, NotNullWhenTrue = 2 }
```

cston replied:
Should the two attributes be combined into a single `[Ensures(...)]` attribute?

jcouv replied:
From our discussion, I'll keep separate attributes for the moment. We'll discuss with LDM.

---
In reply to: [187089388](https://github.com/dotnet/roslyn/pull/26656#discussion_r187089388) [](ancestors = 187089388)

cston commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs:
```diff
@@ -645,6 +647,38 @@ internal override bool IsCallerMemberName
+                    value = _packedFlags.SetWellKnownAttribute(flag,
+                        _moduleSymbol.Module.HasAttribute(_handle, AttributeDescription.NotNullWhenFalseAttribute));
+                }
+                return value || HasExtraAttribute(AttributeDescription.NotNullWhenFalseAttribute);
```
Should `HasExtraAttribute(...)` be included in the `_packedFlags` value, to avoid calling `HasExtraAttribute` each time?

cston commented on src/Compilers/CSharp/Portable/Symbols/ParameterSymbol.cs:
```diff
@@ -385,6 +385,56 @@ internal sealed override ObsoleteAttributeData ObsoleteAttributeData
+            // index 0 is used for return type
+            // `this` parameter is at index 1
+            // other parameters follow
+            int index = this.Ordinal + (containingMethod.IsExtensionMethod ? 2 : 1);
```
Isn't the first parameter for an extension method at the same index as the first parameter of any static method?

jcouv replied:
I couldn't find a good example of a BCL extension method that would need extra annotations. I'll block this scenario for now.

---
In reply to: [187092735](https://github.com/dotnet/roslyn/pull/26656#discussion_r187092735) [](ancestors = 187092735)

cston commented on src/Compilers/CSharp/Portable/Symbols/ParameterSymbol.cs:
```diff
@@ -385,6 +385,56 @@ internal sealed override ObsoleteAttributeData ObsoleteAttributeData
+                return false;
+            }
+
+            string key = ExtraAnnotations.MakeMethodKey(containingMethod);
```
Should `ExtraAnnotations` be an instance, perhaps stored on `CSharpCompilationOptions`? That would allow specific testing in particular.

jcouv replied:
Yes, that seems useful. I'll keep this for follow-up PR though.

---
In reply to: [187120948](https://github.com/dotnet/roslyn/pull/26656#discussion_r187120948) [](ancestors = 187120948)

jcouv commented on src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:
```diff
@@ -5437,6 +5462,549 @@ class CL1<T>
+        }
+
+        [Fact]
+        public void TrueWhenNotNull_String_NotIsNullOrEmpty_NoDuplicateWarnings()
```
The second processing of the invocation (honoring annotations) should not produce duplicate warning.
Without such suppression, we'd get `CS8625` twice on `M2(null)`.

---
In reply to: [186469446](https://github.com/dotnet/roslyn/pull/26656#discussion_r186469446) [](ancestors = 186469446)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1658,22 +1718,134 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
+
+            // Since we may be splitting states to represent the method returning true/false,
+            // we'll prepare all the slots ahead of time.
+            var slots = ArrayBuilder<int?>.GetInstance(arguments.Length);
```
>slots [](start = 16, length = 5)

Why calculate `slots` ahead of time?

jcouv replied:
From our discussion, I'll add a PROTOTYPE comment.

---
In reply to: [187175420](https://github.com/dotnet/roslyn/pull/26656#discussion_r187175420) [](ancestors = 187175420)

cston commented on src/Compilers/CSharp/Portable/Symbols/ExtraAnnotations.cs:
```diff
@@ -0,0 +1,178 @@
+            // PROTOTYPE(NullableReferenceTypes): Many cases are not yet handled
+            // generic type args
+            // ref kind
+            // 'this'
```
>'this' [](start = 15, length = 6)

Does this represent an extension method `this` parameter? Is that important?

jcouv replied:
I'm not sure if that'll be important. I just wanted to keep some notes of things the test if/when we generalize the external annotations.

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/ThisParameterSymbol.cs:
```diff
@@ -114,6 +114,16 @@ internal override bool IsCallerMemberName
 
+        internal override bool NotNullWhenFalse
+        {
+            get { return HasExtraAttribute(AttributeDescription.NotNullWhenFalseAttribute); }
```
>HasExtraAttribute [](start = 25, length = 17)

Will an implicit `this` parameter have attributes?

sharwell commented:
Many code bases already document these conditions using attributes which differ from the ones in this pull request.

### NotNullWhenFalse

:memo: For reference, this:

```
static bool IsNullOrEmpty([NotNullWhenFalse] string? s);
```

is equivalent to this:

```csharp
[Contract("null=>true")]
static bool IsNullOrEmpty(string? s);
```

### EnsuresNotNull

This:

```
static void AssertNotNull<T>([EnsuresNotNull] T? t) where T : class;
```

is equivalent to this:

```
[Contract("t:null=>halt")]
static void AssertNotNull<T>(T? t) where T : class;
```

### Debug.Assert

This:

```
static void Assert([EnsuresTrue] bool condition, string message);
```

is equivalent to this:

```
[AssertionMethod]
static void Assert([AssertionCondition] bool condition, string message);
```

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:
```diff
@@ -5437,6 +5462,1076 @@ class CL1<T>
+{
+    void Main(string? s)
+    {
+        ThrowIfNull(s);
```
>ThrowIfNull(s); [](start = 8, length = 15)

Please add a test for `ThrowIfNull(t); t.ToString();` where t is an unconstrained type parameter.

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:
```diff
@@ -5437,6 +5462,1076 @@ class CL1<T>
+
+            // Also verify from metadata
+            var compilation2 = CreateCompilation("", references: new[] { compilation.EmitToImageReference() });
+            VerifyAnnotationsInSource(compilation2, memberName, expected);
```
Consider renaming `VerifyAnnotationsInSource` since it is not just for source.

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:
```diff
@@ -5437,6 +5462,1076 @@ class CL1<T>
+" + EnsuresNotNullAttributeDefinition, parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics();
+        }
```
Consider testing:
```
static void EnsuresNotNull([EnsuresNotNull] params object[] args) { }
static void F(object? x, object? y, object[]? a)
{
    EnsuresNotNull(a);
    a.ToString(); // no warning
    EnsuresNotNull(x, y);
    x.ToString(); // warning
    y.ToString(); // warning
}

cston replied:
Consider testing:
```
interface I
{
    void F([EnsuresNotNull] object? o);
}
delegate void D([EnsuresNotNull] object? o);
```

jcouv replied:
Good catch. There is a bug on params (the two warnings aren't produced).

---
In reply to: [187462263](https://github.com/dotnet/roslyn/pull/26656#discussion_r187462263) [](ancestors = 187462263)

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:
```diff
@@ -5437,6 +5462,1076 @@ class CL1<T>
+        }
+        else
+        {
+            s.ToString(); // ok
```
>// ok [](start = 26, length = 5)

Minor point: comment is misleading.

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:
```diff
@@ -5437,6 +5462,1076 @@ class CL1<T>
+            s.ToString(); // ok
+        }
+    }
+    public dynamic MyIsNullOrEmpty([NotNullWhenFalse] string? s) => throw null;
```
>MyIsNullOrEmpty [](start = 19, length = 15)

Method is not invoked.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1658,22 +1736,116 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
                 }
-                builder.Add(_result);
             }
-            _result = _invalidType;
```
>_result = _invalidType; [](start = 12, length = 23)

Should `_result = _invalidType;` be added back?

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -3555,6 +3730,21 @@ public bool Reachable
+                }
+
+                return pooledBuilder.ToStringAndFree();
+            }
```
Consider removing the space between each `!?.` so the string representation aligns with the display of the two `BitVector` instances in the debugger.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -3555,6 +3730,21 @@ public bool Reachable
                 }
             }
+
+#if DEBUG
```
Consider including `ToString` regardless of build.

cston commented on src/Compilers/CSharp/Portable/Symbols/ExtraAnnotations.cs:
```diff
@@ -0,0 +1,201 @@
+        internal static string MakeMethodKey(MethodSymbol method)
+        {
+            var containingType = method.ContainingSymbol as TypeSymbol;
+            if (containingType is null)
```
When is `method.ContainingSymbol` not a `TypeSymbol`?

jcouv replied:
I think for local functions in can be a method symbol.

---
In reply to: [187470794](https://github.com/dotnet/roslyn/pull/26656#discussion_r187470794) [](ancestors = 187470794)

cston commented on src/Compilers/CSharp/Portable/Symbols/ExtraAnnotations.cs:
```diff
@@ -0,0 +1,201 @@
+        /// index 0 is used for return type
+        /// other parameters follow
+        /// </summary>
+        internal static (bool hasAny, AttributeAnnotations annotations) GetExtraAttributes(string key, int parameterIndex)
```
Perhaps return `AttributeAnnotations?` instead.

cston commented on src/Compilers/CSharp/Portable/Symbols/ParameterSymbol.cs:
```diff
@@ -385,6 +385,22 @@ internal sealed override ObsoleteAttributeData ObsoleteAttributeData
 
+        internal abstract AttributeAnnotations FlowAnalysisAnnotations { get; }
+
+        protected (bool memberHasExtra, AttributeAnnotations annotations) TryGetExtraAttributeAnnotations()
```
Perhaps return `AttributeAnnotations?`.

cston commented on src/Compilers/CSharp/Portable/Symbols/ParameterSymbol.cs:
```diff
@@ -385,6 +385,22 @@ internal sealed override ObsoleteAttributeData ObsoleteAttributeData
+            ParameterSymbol originalParameter = this.OriginalDefinition;
+            var containingMethod = originalParameter.ContainingSymbol as MethodSymbol;
+
+            if (containingMethod is null)
```
Can we annotate parameters of properties?

jcouv replied:
No, we currently only support extra annotations on methods.

---
In reply to: [187472420](https://github.com/dotnet/roslyn/pull/26656#discussion_r187472420) [](ancestors = 187472420)

jcouv commented on src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:
```diff
@@ -5437,6 +5462,1076 @@ class CL1<T>
+        [Fact]
+        public void NotNullWhenFalse_ReturningObject_FromMetadata()
+        {
+            string il = @"
```
The intent is to validate the loading of misused in metadata. But that metadata can't be compiled because we produce an error for misused attribute in source.

---
In reply to: [187463607](https://github.com/dotnet/roslyn/pull/26656#discussion_r187463607) [](ancestors = 187463607)

cston commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs:
```diff
@@ -645,6 +647,44 @@ internal override bool IsCallerMemberName
+
+                // PROTOTYPE(NullableReferenceTypes): the flags could be packed more
+                if (!_packedFlags.TryGetWellKnownAttribute(notNullWhenFalse, out bool hasNotNullWhenFalse) ||
+                    !_packedFlags.TryGetWellKnownAttribute(ensuresNotNull, out bool hasEnsuresNotNull))
```
To avoid race conditions, we probably need to get the two bits atomically (and then set the two bits atomically), or get and set the two bits independently.

gafter commented on src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:
```diff
@@ -1609,6 +1609,8 @@ internal enum ErrorCode
         ERR_NotNullableOperatorNotReferenceType = 8624,
         WRN_NullAsNonNullable = 8625,
         WRN_NoBestNullabilityConditionalExpression = 8626,
+        ERR_AttributeRequiresBoolReturn = 8627,
```
>ERR_AttributeRequiresBoolReturn [](start = 8, length = 31)

I didn't notice where these are used, but I'm not confident it is a good idea to have diagnostics such as this.

gafter replied:
Because of "forward compatibility" scenarios. (Code that compiles with older compilers will fail to compile with newer compilers)

---
In reply to: [187484195](https://github.com/dotnet/roslyn/pull/26656#discussion_r187484195) [](ancestors = 187484195)

jcouv replied:
Thanks for the clarification.
Filed https://github.com/dotnet/roslyn/issues/26783 to track compat concern.

gafter commented:
>             bool expanded)

drop unused parameter?

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:1728 in 229d8d0. [](commit_id = 229d8d0b24eceb06e3070c28b55dfb9079e2f1ad, deletion_comment = False)

gafter commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1658,22 +1736,117 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
+                    LocalState whenFalse = this.StateWhenFalse.Clone();
+
+                    this.SetState(whenTrue);
+                    VisitArgumentEvaluate(arguments, refKindsOpt, i);
```
>VisitArgumentEvaluate [](start = 20, length = 21)

You're evaluating even previous arguments here.

jcouv commented:
I've found a bug with `out` argument scenario (which doesn't expect that an invocation could return in a split state). I'll fix in the next PR.

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/Attributes/CommonParameterWellKnownAttributeData.cs:
```diff
@@ -109,5 +109,37 @@ public bool HasIUnknownConstantAttribute
         #endregion
+
+        private bool _hasNotNullWhenFalseAttribute;
+        public bool HasNotNullWhenFalseAttribute
```
>public bool HasNotNullWhenFalseAttribute [](start = 8, length = 40)

I assume VB doesn't take advantage of this member, please move it to C# specific type. Declare one if it doesn't exist.

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/Attributes/CommonParameterWellKnownAttributeData.cs:
```diff
@@ -109,5 +109,37 @@ public bool HasIUnknownConstantAttribute
+        }
+
+        private bool _hasEnsuresNotNullAttribute;
+        public bool HasEnsuresNotNullAttribute
```
>HasEnsuresNotNullAttribute [](start = 20, length = 26)

I assume VB doesn't take advantage of this member, please move it to C# specific type.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/ExtraAnnotations.cs:
```diff
@@ -0,0 +1,202 @@
+    internal enum AttributeAnnotations
+    {
+        None = 0,
+        NotNullWhenFalse,
```
>NotNullWhenFalse [](start = 8, length = 16)

Please use explicit values for ````[Flags]``` enums

jcouv replied:
Done in follow-up PR https://github.com/dotnet/roslyn/pull/26828

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/ExtraAnnotations.cs:
```diff
@@ -0,0 +1,202 @@
+        /// index 0 is used for return type
+        /// other parameters follow
+        /// </summary>
+        internal static (bool hasAny, AttributeAnnotations annotations) GetExtraAttributes(string key, int parameterIndex)
```
>bool hasAny [](start = 25, length = 11)

I am not sure what extra information do we get from ```hasAny``` flag. Would it be sufficient to simply check ````AttributeAnnotations``` for 0?

AlekseyTs replied:
Perhaps a clear comment might help

---
In reply to: [187938913](https://github.com/dotnet/roslyn/pull/26656#discussion_r187938913) [](ancestors = 187938913)

jcouv replied:
Added in follow-up PR https://github.com/dotnet/roslyn/pull/26828/files#diff-d02e2d6d70e84697ca87520824236d04R169

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/ParameterSymbol.cs:
```diff
@@ -385,6 +385,22 @@ internal sealed override ObsoleteAttributeData ObsoleteAttributeData
 
+        internal abstract AttributeAnnotations FlowAnalysisAnnotations { get; }
+
+        protected (bool memberHasExtra, AttributeAnnotations annotations) TryGetExtraAttributeAnnotations()
```
>memberHasExtra [](start = 24, length = 14)

The value of ```AttributeAnnotations``` member is not obvious to me.

AlekseyTs replied:
Perhaps a clear comment might help

---
In reply to: [187939172](https://github.com/dotnet/roslyn/pull/26656#discussion_r187939172) [](ancestors = 187939172)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceClonedParameterSymbol.cs:
```diff
@@ -151,6 +151,11 @@ internal override bool IsCallerMemberName
 
+        internal override AttributeAnnotations FlowAnalysisAnnotations
+        {
+            get { return _originalParam.FlowAnalysisAnnotations; }
```
>_originalParam.FlowAnalysisAnnotations [](start = 25, length = 38)

Do we have a test covering this code path?

jcouv replied:
Add `EnsuresNotNull_BeginInvoke` and adjusted behavior (the attributes don't help nullability analysis on BeginInvoke/EndInvoke).

---
In reply to: [187942517](https://github.com/dotnet/roslyn/pull/26656#discussion_r187942517) [](ancestors = 187942517)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:
```diff
@@ -127,6 +127,24 @@ private bool HasCallerMemberNameAttribute
+                (bool memberHasAny, AttributeAnnotations annotations) = TryGetExtraAttributeAnnotations();
+                if (memberHasAny)
+                {
+                    return annotations;
```
>return annotations [](start = 20, length = 18)

Do we have a test covering this code path?

jcouv replied:
Debug.Assert is actually not in the follow-up PR, but the one after that. Added a tracking comment in the meantime: https://github.com/dotnet/roslyn/pull/26828/files#diff-829bf456227af3ee271c7ae37b62ffc0R7056

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:
```diff
@@ -127,6 +127,24 @@ private bool HasCallerMemberNameAttribute
+
+                CommonParameterWellKnownAttributeData attributeData = GetDecodedWellKnownAttributeData();
+
+                return AttributeAnnotations.None
```
>return AttributeAnnotations.None [](start = 16, length = 32)

Do we have a test covering this code path?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceSimpleParameterSymbol.cs:
```diff
@@ -82,6 +82,15 @@ internal override bool IsCallerMemberName
+            get
+            {
+                (_, AttributeAnnotations annotations) = TryGetExtraAttributeAnnotations();
+                return annotations;
```
>return annotations; [](start = 16, length = 19)

Do we have tests covering this code path?

jcouv replied:
`NotNullWhenFalse_String_IsNullOrEmpty_WithoutCorlib` and `NotNullWhenFalse_String_IsNullOrWhiteSpace_WithoutCorlib` for case with annotations.
Most existing nullability tests for case with no annotations returned.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Wrapped/WrappedParameterSymbol.cs:
```diff
@@ -153,7 +153,12 @@ internal override bool IsCallerMemberName
         }
 
-        public override string GetDocumentationCommentXml(CultureInfo preferredCulture = null, bool expandIncludes = false, CancellationToken cancellationToken = default(CancellationToken))
+        internal override AttributeAnnotations FlowAnalysisAnnotations
```
>internal override AttributeAnnotations FlowAnalysisAnnotations [](start = 8, length = 62)

In general, in symbols only leaf classes should provide implementation and each leaf class should be covered by targeted tests. This implementation is fine for the prototype, but, if we decide to go with this approach, we should follow up. For example, I will not be surprised if retargeting symbols would need special handling, etc. Please add a prototype comment.

AlekseyTs commented on src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:
```diff
@@ -306,19 +306,21 @@ public static SyntaxTree ParseWithRoundTripCheck(string text, CSharpParseOptions
             CSharpCompilationOptions options = null,
-            bool appendDefaultHeader = true) => CreateCompilationWithILAndMscorlib40(source, ilSource, TargetFramework.Standard, references, options, appendDefaultHeader);
+            CSharpParseOptions parseOptions = null,
+            bool appendDefaultHeader = true) => CreateCompilationWithILAndMscorlib40(source, ilSource, TargetFramework.Standard, references, options, parseOptions, appendDefaultHeader);
```
>bool appendDefaultHeader = true) => CreateCompilationWithILAndMscorlib40(source, ilSource, TargetFramework.Standard, references, options, parseOptions, appendDefaultHeader); [](start = 12, length = 173)

Given a number of parameters for the method and the presence of optional ones, the expression body form is "unreadable". Consider converting this method to the regular body form.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1603,6 +1613,49 @@ public override BoundNode VisitCall(BoundCall node)
+
+                // We'll ignore NotNullWhenFalse that is misused in metadata
+                if ((annotations & AttributeAnnotations.NotNullWhenFalse) != 0 &&
+                    parameter.ContainingSymbol.GetTypeOrReturnType().SpecialType != SpecialType.System_Boolean)
```
>parameter [](start = 20, length = 9)

Should we use original definition for this check? Is generic case covered by tests?

jcouv replied:
Added ` NotNullWhenFalse_RequiresBoolReturn_OnGenericMethod`. The behavior of checking the specialized definition, rather than the original definition, seem ok.

---
In reply to: [187948468](https://github.com/dotnet/roslyn/pull/26656#discussion_r187948468) [](ancestors = 187948468)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1603,6 +1613,49 @@ public override BoundNode VisitCall(BoundCall node)
+
+                // We'll ignore EnsuresNotNull that is misused in metadata
+                if ((annotations & AttributeAnnotations.EnsuresNotNull) != 0 &&
+                    (parameter.Type?.IsValueType != false || parameter.IsParams))
```
>parameter [](start = 21, length = 9)

Should we use original definition for this check? Is generic case covered by tests?

jcouv replied:
I'll add tests.
To do that more easily, I'll pull the change to remove the warning for misused attributes in this PR. That'll also address some other feedback.

jcouv replied:
Added tests `EnsuresNotNull_Generic_WithRefType`, `EnsuresNotNull_Generic_WithValueType` and `EnsuresNotNull_Generic_WithUnconstrainedGenericType`.

---
In reply to: [187948601](https://github.com/dotnet/roslyn/pull/26656#discussion_r187948601) [](ancestors = 187948601)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:
```diff
@@ -605,6 +623,28 @@ internal override void DecodeWellKnownAttribute(ref DecodeWellKnownAttributeArgu
+            }
+            else if (attribute.IsTargetAttribute(this, AttributeDescription.EnsuresNotNullAttribute))
+            {
+                if (this.Type.IsValueType)
```
>this.Type.IsValueType [](start = 20, length = 21)

What if type is an unconstraint generic type parameter?

jcouv replied:
Unconstrained generic type gives `IsValueType` as `false` and no warning is produced.
That said, I'm removing those warnings in follow-up PR, from discussion with Neal and Chuck (those warning create compat issue once the attributes are released, if someone uses the attribute with C# 7 compiler then upgrades to C# 8 compiler).

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:
```diff
@@ -605,6 +623,28 @@ internal override void DecodeWellKnownAttribute(ref DecodeWellKnownAttributeArgu
             }
+            else if (attribute.IsTargetAttribute(this, AttributeDescription.NotNullWhenFalseAttribute))
+            {
+                if (this.ContainingSymbol.GetTypeOrReturnType().SpecialType == SpecialType.System_Boolean)
```
>this.ContainingSymbol.GetTypeOrReturnType() [](start = 20, length = 43)

What if type is an unconstraint generic type parameter?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1684,22 +1761,117 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
+                else if (notNullWhenFalse)
+                {
+                    // We'll use the WhenTrue/False states to represent whether the invocation returns true/false
+                    Split();
```
>Split(); [](start = 20, length = 8)

I think think this might be a wrong place to do the Split, it should be done when we are visiting invocation, not the arguments.

jcouv replied:
I don't see a functional advantage to splitting early. Do you mean that would be better for clarity?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -3573,6 +3748,19 @@ public bool Reachable
                 }
             }
+
+            public override string ToString()
```
>ToString [](start = 35, length = 8)

Is this for debugging purposes? Consider adding a specialized helper for that rather than to changing ToString.

jcouv replied:
I've changed that to use debugger display method instead, in follow-up PR https://github.com/dotnet/roslyn/pull/26828

AlekseyTs commented:
Done with review pass (iteration 13)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1684,22 +1761,117 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
+                    }
+                    else
+                    {
+                        this.State[slot] = true;
```
>this.State[slot] = true [](start = 24, length = 23)

It feels like in general case prior arguments should be able to take advantage of this information.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1684,22 +1761,117 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
+                    }
+                    else
+                    {
+                        this.State[slot] = true;
```
>this.State[slot] = true; [](start = 24, length = 24)

It feels like in general case this information should flow into type argument inference for the method, but it doesn't look like it does.

jcouv replied:
Added tests `EnsuresNotNull_BackEffect`, `EnsuresNotNull_ForwardEffect`(s)

---
In reply to: [188022000](https://github.com/dotnet/roslyn/pull/26656#discussion_r188022000) [](ancestors = 188022000)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:
```diff
@@ -5437,6 +5462,1262 @@ class CL1<T>
+{
+    public void Main(string? s)
+    {
+        if (MyIsNullOrEmpty(s))
```
>MyIsNullOrEmpty(s) [](start = 12, length = 18)

Please add a test for conditions of the form ```MyIsNullOrEmpty(s) == true``` and ```MyIsNullOrEmpty(s) == false```. And also with comparison operands in different order.

jcouv replied:
That is indeed broken. Thanks!
Add tests `NotNullWhenFalse_ComparedToTrue` and `NotNullWhenFalse_ComparedToFalse`

---
In reply to: [188023547](https://github.com/dotnet/roslyn/pull/26656#discussion_r188023547) [](ancestors = 188023547)

jcouv commented:
@AlekseyTs This PR is ready for another look. Thanks
